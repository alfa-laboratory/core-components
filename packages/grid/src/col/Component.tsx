import cn from 'classnames';
import React, { useMemo } from 'react';
import { ResponsivePropertyType } from '../typings';
import { createClassNames } from '../utils';

import styles from './index.module.css';
import guttersStyles from '../gutters.module.css';

export type ColProps = {
    /**
     * Дополнительный класс
     */
    className?: string;

    /**
     * Управление выравниванием колонок по вертикальной оси
     */
    align?: 'top' | 'middle' | 'bottom';

    /**
     * Управление шириной колонки.
     * Возможные значения: `[1..12, available, auto]`
     * или `{ mobile: [1..12], tablet: [1..12], desktop: [1..12] }`
     * или `{ mobile: { s: [1..12], m: [1..12], l: [1..12] },
     * tablet: { s: [1..12], m: [1..12] },
     * desktop: { s: [1..12], m: [1..12], l: [1..12], xl: [1..12] } }`.
     */
    width?: ResponsivePropertyType;

    /**
     * Управлние смещением колонки.
     * Возможные значения: `[1..11]`
     * или `{ mobile: [1..11], tablet: [1..11], desktop: [1..11] }`
     * или `{ mobile: { s: [1..11], m: [1..11], l: [1..11] },
     * tablet: { s: [1..11], m: [1..11] },
     * desktop: { s: [1..11], m: [1..11], l: [1..11], xl: [1..11] } }`.
     */
    offset?: ResponsivePropertyType;

    /**
     * Управление порядком колонок.
     * Возможные значения: `[1..12, first, last]`
     * или `{ mobile: [1..last], tablet: [1..last], desktop: [1..last] }`
     * или `{ mobile: { s: [1..last], m: [1..last], l: [1..last] },
     * tablet: { s: [1..last], m: [1..last] },
     * desktop: { s: [1..last], m: [1..last], l: [1..last], xl: [1..last] } }`.
     */
    order?: ResponsivePropertyType;

    /**
     * Html тег компонента.
     */
    tag?: keyof JSX.IntrinsicElements;

    /**
     * Контент
     */
    children?: React.ReactNode;

    /**
     * Идентификатор для систем автоматизированного тестирования
     */
    dataTestId?: string;
};

export const Col: React.FC<ColProps> = ({
    tag: Component = 'div',
    className,
    align,
    order,
    offset,
    width,
    children,
    dataTestId,
}) => {
    const classNames = useMemo(() => createClassNames({ order, offset, width }, styles), [
        order,
        offset,
        width,
    ]);

    return (
        <Component
            className={cn(
                guttersStyles.col,
                styles.component,
                align && styles[`align-${align}`],
                ...classNames,
                className,
            )}
            data-test-id={dataTestId}
        >
            {children}
        </Component>
    );
};
