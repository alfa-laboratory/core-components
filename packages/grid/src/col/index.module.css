:root {
    --grid-col-width: calc(100% / 12);
}

.component {
    max-width: 100%;
    min-height: 1px;
    flex-basis: 0;
    flex-grow: 1;
    box-sizing: border-box;
}

.width-auto {
    flex: 0 0 auto;
    width: auto;
    max-width: none;
}

.width-available {
    flex-basis: auto;
    min-height: auto;
    height: auto;
}

@for $i from 1 to 12 {
    .width-$(i) {
        flex: 0 0 calc(var(--grid-col-width) * $(i));
        max-width: calc(var(--grid-col-width) * $(i));
    }
}

.align-top {
    align-self: flex-start;
}

.align-middle {
    align-self: center;
}

.align-bottom {
    align-self: flex-end;
}

@for $i from 1 to 11 {
    .offset-$(i) {
        margin-left: calc(var(--grid-col-width) * $(i));
    }
}

.order-first {
    order: -1;
}

.order-last {
    order: 13;
}

@for $i from 1 to 12 {
    .order-$(i) {
        order: $(i);
    }
}

@each $breakpoint in mobile, mobile-s, mobile-m, mobile-l, tablet, tablet-s, tablet-m, desktop,
    desktop-s, desktop-m, desktop-l, desktop-xl
{
    @media (--$(breakpoint)) {
        .width-$(breakpoint)-auto {
            flex: 0 0 auto;
            width: auto;
            max-width: none;
        }

        .width-$(breakpoint)-available {
            flex-basis: auto;
            min-height: auto;
            height: auto;
        }

        @for $i from 1 to 12 {
            .width-$(breakpoint)-$(i) {
                flex: 0 0 calc(var(--grid-col-width) * $(i));
                max-width: calc(var(--grid-col-width) * $(i));
            }
        }

        @for $i from 1 to 11 {
            .offset-$(breakpoint)-$(i) {
                margin-left: calc(var(--grid-col-width) * $(i));
            }
        }

        .order-$(breakpoint)-first {
            order: -1;
        }

        .order-$(breakpoint)-last {
            order: 13;
        }

        @for $i from 1 to 12 {
            .order-$(breakpoint)-$(i) {
                order: $(i);
            }
        }
    }
}
