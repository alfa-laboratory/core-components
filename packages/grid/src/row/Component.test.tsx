import React from 'react';
import { render } from '@testing-library/react';

import { Col } from '../col/index';
import { Row } from './index';

describe('Grid.Row', () => {
    describe('Snapshots tests', () => {
        it('should match snapshot', () => {
            const { container } = render(
                <Row>
                    <Col>content</Col>
                </Row>,
            );

            expect(container).toMatchSnapshot();
        });
    });

    describe('Classes tests', () => {
        it('should set custom class', () => {
            const className = 'custom-class';

            const { container } = render(<Row className={className} />);

            expect(container.firstElementChild).toHaveClass(className);
        });

        it('should set gutter class', () => {
            const { container } = render(
                <Row gutter={{ mobile: 0, tablet: 8, desktop: { s: 16, l: 24 } }} />,
            );

            expect(container.firstElementChild).toHaveClass('gutter-mobile-0');
            expect(container.firstElementChild).toHaveClass('gutter-tablet-8');
            expect(container.firstElementChild).toHaveClass('gutter-desktop-s-16');
            expect(container.firstElementChild).toHaveClass('gutter-desktop-l-24');
        });

        it('should set align class', () => {
            const { container } = render(<Row align='top' />);

            expect(container.firstElementChild).toHaveClass('top');
        });

        it('should set justify class', () => {
            const { container } = render(<Row justify='center' />);

            expect(container.firstElementChild).toHaveClass('center');
        });
    });

    it('should set `data-test-id` attribute to row', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Row dataTestId={dataTestId} />);

        expect(getByTestId(dataTestId).tagName).toBe('DIV');
    });

    it('should set tag correcly', () => {
        const dataTestId = 'test-id';
        const { getByTestId } = render(<Row tag='section' dataTestId={dataTestId} />);

        expect(getByTestId(dataTestId).tagName).toBe('SECTION');
    });
});
