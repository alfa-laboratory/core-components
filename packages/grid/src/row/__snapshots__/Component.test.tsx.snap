// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid.Row Snapshots tests should match snapshot 1`] = `
<div>
  <div
    class="row component between gutter-mobile-s-16 gutter-desktop-m-24"
  >
    <div
      class="col component"
    >
      content
    </div>
  </div>
</div>
`;
